File: services/crm-sync-service/internal/metrics/metrics.go
// Package metrics is a tiny dependency-free Prometheus text-format registry:
// counters per event type and Twenty call latency summaries (SPEC-CRM §B).
package metrics

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Registry collects counters, latency summaries and the Twenty-call
// histogram.
type Registry struct {
	mu         sync.Mutex
	counters   map[string]int64
	latency    map[string]*latStat
	histograms map[string]*histogram
}

type latStat struct {
	count int64
	sum   float64 // seconds
	max   float64
}

// New builds an empty registry.
func New() *Registry {
	return &Registry{
		counters:   map[string]int64{},
		latency:    map[string]*latStat{},
		histograms: map[string]*histogram{},
	}
}

// histogram is a fixed-bucket cumulative histogram.
type histogram struct {
	buckets []float64 // upper bounds, sorted, without +Inf
	counts  []int64   // len(buckets)+1; counts[i] = observations <= buckets[i] (cumulative)
	sum     float64
	count   int64
}

// twentyCallBuckets: Twenty REST calls are local-network; sub-10s range.
var twentyCallBuckets = []float64{0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10}

func newHistogram(buckets []float64) *histogram {
	return &histogram{buckets: buckets, counts: make([]int64, len(buckets)+1)}
}

func (h *histogram) observe(sec float64) {
	h.count++
	h.sum += sec
	for i, b := range h.buckets {
		if sec <= b {
			h.counts[i]++
		}
	}
	h.counts[len(h.buckets)]++ // +Inf bucket
}

// ObserveTwentyCall records one Twenty REST call duration (seconds) in the
// crm_sync_twenty_call_duration_seconds histogram, labelled by HTTP method
// and path class (the Twenty object, e.g. "people", "tasks").
func (r *Registry) ObserveTwentyCall(method, pathClass string, d time.Duration) {
	r.mu.Lock()
	defer r.mu.Unlock()
	key := method + "|" + pathClass
	h := r.histograms[key]
	if h == nil {
		h = newHistogram(twentyCallBuckets)
		r.histograms[key] = h
	}
	h.observe(d.Seconds())
}

// Inc increments counter `name` by 1.
func (r *Registry) Inc(name string) { r.Add(name, 1) }

// Add increments counter `name` by delta.
func (r *Registry) Add(name string, delta int64) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.counters[name] += delta
}

// Observe records one latency observation (seconds) under `name`.
func (r *Registry) Observe(name string, d time.Duration) {
	r.mu.Lock()
	defer r.mu.Unlock()
	s := r.latency[name]
	if s == nil {
		s = &latStat{}
		r.latency[name] = s
	}
	sec := d.Seconds()
	s.count++
	s.sum += sec
	if sec > s.max {
		s.max = sec
	}
}

// Render emits the Prometheus text exposition format.
func (r *Registry) Render() string {
	r.mu.Lock()
	defer r.mu.Unlock()
	var b strings.Builder
	keys := make([]string, 0, len(r.counters))
	for k := range r.counters {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	b.WriteString("# HELP crm_sync_counter Counters by name (events processed/failed/DLQ, webhook intakes).\n")
	b.WriteString("# TYPE crm_sync_counter counter\n")
	for _, k := range keys {
		fmt.Fprintf(&b, "crm_sync_counter{name=%q} %d\n", k, r.counters[k])
	}
	lkeys := make([]string, 0, len(r.latency))
	for k := range r.latency {
		lkeys = append(lkeys, k)
	}
	sort.Strings(lkeys)
	b.WriteString("# HELP crm_sync_latency_seconds Latency summary in seconds by operation.\n")
	b.WriteString("# TYPE crm_sync_latency_seconds summary\n")
	for _, k := range lkeys {
		s := r.latency[k]
		fmt.Fprintf(&b, "crm_sync_latency_seconds_count{op=%q} %d\n", k, s.count)
		fmt.Fprintf(&b, "crm_sync_latency_seconds_sum{op=%q} %.6f\n", k, s.sum)
		fmt.Fprintf(&b, "crm_sync_latency_seconds_max{op=%q} %.6f\n", k, s.max)
	}
	hkeys := make([]string, 0, len(r.histograms))
	for k := range r.histograms {
		hkeys = append(hkeys, k)
	}
	sort.Strings(hkeys)
	b.WriteString("# HELP crm_sync_twenty_call_duration_seconds Twenty REST call duration histogram.\n")
	b.WriteString("# TYPE crm_sync_twenty_call_duration_seconds histogram\n")
	for _, k := range hkeys {
		method, pathClass, _ := strings.Cut(k, "|")
		h := r.histograms[k]
		for i, bound := range h.buckets {
			fmt.Fprintf(&b, "crm_sync_twenty_call_duration_seconds_bucket{method=%q,path_class=%q,le=%q} %d\n",
				method, pathClass, strconv.FormatFloat(bound, 'g', -1, 64), h.counts[i])
		}
		fmt.Fprintf(&b, "crm_sync_twenty_call_duration_seconds_bucket{method=%q,path_class=%q,le=\"+Inf\"} %d\n",
			method, pathClass, h.counts[len(h.buckets)])
		fmt.Fprintf(&b, "crm_sync_twenty_call_duration_seconds_sum{method=%q,path_class=%q} %.6f\n",
			method, pathClass, h.sum)
		fmt.Fprintf(&b, "crm_sync_twenty_call_duration_seconds_count{method=%q,path_class=%q} %d\n",
			method, pathClass, h.count)
	}
	return b.String()
}

File: services/voice-agent-runtime/app/metrics.py
"""Hand-rolled Prometheus text exposition for the voice inference plane
(VOICE-SCALING §3): per-call STT/LLM/TTS latency, LLM token usage, tool-call
outcomes and active voice sessions, served on the control plane `/metrics`.

No prometheus-client dependency: the registry is a small thread-safe in-memory
store with a `render()` producing the standard text exposition format, so the
worker job processes and the control plane can both record at low overhead.

Series:
- voice_stt_latency_seconds   (histogram)
- voice_llm_latency_seconds   (histogram)
- voice_llm_tokens_total      (counter, label kind=prompt|completion)
- voice_tts_latency_seconds   (histogram)
- voice_tool_calls_total      (counter, labels tool,result)
- voice_active_sessions       (gauge)
"""

from __future__ import annotations

import contextvars
import threading
import time
from typing import Any

DEFAULT_BUCKETS = (0.05, 0.1, 0.25, 0.5, 1.0, 2.0, 4.0, 8.0)


def _labels_key(labels: dict[str, str]) -> tuple[tuple[str, str], ...]:
    return tuple(sorted(labels.items()))


def _format_labels(labels: tuple[tuple[str, str], ...]) -> str:
    if not labels:
        return ""
    inner = ",".join(f'{k}="{v}"' for k, v in labels)
    return "{" + inner + "}"


class Histogram:
    def __init__(self, name: str, description: str, buckets=DEFAULT_BUCKETS) -> None:
        self.name = name
        self.description = description
        self.buckets = tuple(sorted(buckets))
        self._counts = [0] * len(self.buckets)
        self._sum = 0.0
        self._count = 0
        self._lock = threading.Lock()

    def observe(self, value: float) -> None:
        with self._lock:
            for i, bound in enumerate(self.buckets):
                if value <= bound:
                    self._counts[i] += 1
            self._sum += value
            self._count += 1

    def time(self) -> "_Timer":
        return _Timer(self)

    def render(self) -> list[str]:
        lines = [
            f"# HELP {self.name} {self.description}",
            f"# TYPE {self.name} histogram",
        ]
        with self._lock:
            counts = list(self._counts)
            total = self._count
            sum_ = self._sum
        for bound, count in zip(self.buckets, counts):
            lines.append(f'{self.name}_bucket{{le="{bound:g}"}} {count}')
        lines.append(f'{self.name}_bucket{{le="+Inf"}} {total}')
        lines.append(f"{self.name}_sum {sum_:.6f}")
        lines.append(f"{self.name}_count {total}")
        return lines


class _Timer:
    """Context manager: `with hist.time(): ...` records wall-clock seconds."""

    def __init__(self, hist: Histogram) -> None:
        self._hist = hist
        self._start = 0.0
        self.elapsed = 0.0  # seconds measured by the last `with` block

    def __enter__(self) -> "_Timer":
        self._start = time.perf_counter()
        return self

    def __exit__(self, *exc) -> None:
        self.elapsed = time.perf_counter() - self._start
        self._hist.observe(self.elapsed)


class Counter:
    def __init__(self, name: str, description: str, label_names: tuple[str, ...]) -> None:
        self.name = name
        self.description = description
        self.label_names = label_names
        self._values: dict[tuple[tuple[str, str], ...], float] = {}
        self._lock = threading.Lock()

    def inc(self, amount: float = 1.0, **labels: str) -> None:
        key = _labels_key({k: str(v) for k, v in labels.items()})
        with self._lock:
            self._values[key] = self._values.get(key, 0.0) + amount

    def render(self) -> list[str]:
        lines = [
            f"# HELP {self.name} {self.description}",
            f"# TYPE {self.name} counter",
        ]
        with self._lock:
            items = sorted(self._values.items())
        for key, value in items:
            lines.append(f"{self.name}{_format_labels(key)} {value:g}")
        return lines


class Gauge:
    def __init__(self, name: str, description: str) -> None:
        self.name = name
        self.description = description
        self._value = 0.0
        self._lock = threading.Lock()

    def set(self, value: float) -> None:
        with self._lock:
            self._value = value

    def inc(self, amount: float = 1.0) -> None:
        with self._lock:
            self._value += amount

    def dec(self, amount: float = 1.0) -> None:
        with self._lock:
            self._value -= amount

    def render(self) -> list[str]:
        with self._lock:
            value = self._value
        return [
            f"# HELP {self.name} {self.description}",
            f"# TYPE {self.name} gauge",
            f"{self.name} {value:g}",
        ]


class Registry:
    def __init__(self) -> None:
        self.stt_latency = Histogram(
            "voice_stt_latency_seconds", "STT transcription latency per call."
        )
        self.llm_latency = Histogram(
            "voice_llm_latency_seconds", "LLM chat-completion latency per call."
        )
        self.llm_tokens = Counter(
            "voice_llm_tokens_total",
            "LLM tokens consumed.",
            ("kind",),  # kind=prompt|completion
        )
        self.tts_latency = Histogram(
            "voice_tts_latency_seconds", "TTS synthesis latency per call."
        )
        self.tool_calls = Counter(
            "voice_tool_calls_total",
            "Tool calls executed by the voice tool layer.",
            ("tool", "result"),  # result=ok|error|timeout
        )
        self.active_sessions = Gauge(
            "voice_active_sessions", "Currently active voice/chat sessions."
        )

    def render(self) -> str:
        lines: list[str] = []
        for metric in (
            self.stt_latency,
            self.llm_latency,
            self.llm_tokens,
            self.tts_latency,
            self.tool_calls,
            self.active_sessions,
        ):
            lines.extend(metric.render())
        return "\n".join(lines) + "\n"


# Process-wide default registry. Call sites record directly into it; tests
# swap it via `set_registry`.
_registry = Registry()


def get_registry() -> Registry:
    return _registry


def set_registry(registry: Registry) -> None:
    global _registry
    _registry = registry


def reset_registry() -> Registry:
    """Fresh registry (test isolation); returns it."""
    registry = Registry()
    set_registry(registry)
    return registry


def render() -> str:
    return _registry.render()


# ---------------------------------------------------------------------------
# Per-session call-quality accumulator (CRM gap fix): while the registry
# above aggregates process-wide for Prometheus, a SessionMetrics instance
# tracks the signals of ONE conversation so they can be attached to the
# SessionEnded CloudEvent (consumed by crm-sync-service to write a Twenty
# call-summary note).
#
# The "active session" is a contextvar: the LiveKit worker activates one
# SessionMetrics per job (one room = one session per job process), and the
# recording call sites below no-op when no session is active (control-plane
# scrape paths, unit tests, sessions without instrumentation).
# ---------------------------------------------------------------------------
class SessionMetrics:
    """Thread-safe accumulator for one conversation's quality signals.

    Records STT/TTS call counts, per-call LLM latencies, tool-call counts,
    turn count and whether the LLM fallback chain was used. `quality_payload`
    renders the `quality` object for the SessionEnded event, or None when
    nothing was recorded (guard: no data -> no quality key on the event).
    """

    def __init__(self, conversation_id: str, *, clock=time.time) -> None:
        self.conversation_id = conversation_id
        self._clock = clock
        self.started_at = clock()
        self.turn_count = 0
        self.stt_calls = 0
        self.tts_calls = 0
        self.llm_fallback_used = False
        self._llm_latencies_ms: list[float] = []
        self._tool_calls: dict[str, int] = {}
        self._lock = threading.Lock()

    # ------------------------------------------------------------- recording
    def record_turn(self) -> None:
        with self._lock:
            self.turn_count += 1

    def record_stt(self) -> None:
        with self._lock:
            self.stt_calls += 1

    def record_tts(self) -> None:
        with self._lock:
            self.tts_calls += 1

    def record_tool_call(self, name: str) -> None:
        with self._lock:
            self._tool_calls[name] = self._tool_calls.get(name, 0) + 1

    def record_llm_latency(self, seconds: float) -> None:
        with self._lock:
            self._llm_latencies_ms.append(seconds * 1000.0)

    def record_llm_fallback(self) -> None:
        with self._lock:
            self.llm_fallback_used = True

    # -------------------------------------------------------------- snapshot
    def tool_calls(self) -> dict[str, int]:
        with self._lock:
            return dict(self._tool_calls)

    def has_data(self) -> bool:
        with self._lock:
            return bool(
                self.turn_count
                or self.stt_calls
                or self.tts_calls
                or self._llm_latencies_ms
                or self._tool_calls
            )

    def quality_payload(
        self,
        *,
        escalated: bool = False,
        confirmed_phone: str | None = None,
    ) -> dict[str, Any] | None:
        """SessionEnded `quality` object, or None when nothing was recorded.

        Latency fields are null (not 0) when the session made no LLM calls
        through the instrumented path (e.g. the LiveKit worker's plugin LLM
        node — only the fallback chain and the chat tool loop time calls).
        """
        with self._lock:
            latencies = list(self._llm_latencies_ms)
            tools = dict(self._tool_calls)
            empty = not (
                self.turn_count or self.stt_calls or self.tts_calls or latencies or tools
            )
            duration_s = round(self._clock() - self.started_at, 1)
            turn_count = self.turn_count
            stt_calls = self.stt_calls
            tts_calls = self.tts_calls
            fallback = self.llm_fallback_used
        if empty:
            return None
        return {
            "duration_s": duration_s,
            "turn_count": turn_count,
            "tool_calls": tools,
            "avg_llm_latency_ms": (
                round(sum(latencies) / len(latencies)) if latencies else None
            ),
            "max_llm_latency_ms": round(max(latencies)) if latencies else None,
            "stt_calls": stt_calls,
            "tts_calls": tts_calls,
            "llm_fallback_used": fallback,
            "escalated": escalated,
            "confirmed_phone": confirmed_phone,
        }


_active_session: contextvars.ContextVar[SessionMetrics | None] = contextvars.ContextVar(
    "voice_session_metrics", default=None
)


def activate_session(session: SessionMetrics) -> SessionMetrics:
    """Make `session` the active per-session accumulator for this context.

    Returns the session so callers can keep a direct reference. The binding
    lives for the rest of the context (a LiveKit job runs exactly one
    session), so no explicit deactivate is needed; tests can rebind freely.
    """
    _active_session.set(session)
    return session


def get_active_session() -> SessionMetrics | None:
    return _active_session.get()


# Recording helpers used at the same call sites as the global registry. All
# are no-ops when no session is active in this context.
def session_turn() -> None:
    if (s := get_active_session()) is not None:
        s.record_turn()


def session_stt() -> None:
    if (s := get_active_session()) is not None:
        s.record_stt()


def session_tts() -> None:
    if (s := get_active_session()) is not None:
        s.record_tts()


def session_tool_call(name: str) -> None:
    if (s := get_active_session()) is not None:
        s.record_tool_call(name)


def session_llm_latency(seconds: float) -> None:
    if (s := get_active_session()) is not None:
        s.record_llm_latency(seconds)


def session_llm_fallback() -> None:
    if (s := get_active_session()) is not None:
        s.record_llm_fallback()
